modular_bandastation/automapper/code/area_spawn_subsystem.dm

```diff
@@ -0,0 +1,376 @@
+			continue
+
+		// Every other mode.
+		if(
+			found_movable.density \
+			|| is_type_in_list(found_movable, restricted_objects_list) \
+			|| is_type_in_list(found_movable, restricted_overlap_objects_list)
+		)
+			return 0
+
+		if(found_movable.layer > LOW_OBJ_LAYER && found_movable.layer < ABOVE_MOB_LAYER)
+			totally_empty = FALSE
+
+	// Number of directions that have a closed wall
+	var/num_walls_found = 0
+	// Found a dense object?
+	var/found_dense_object = FALSE
+	// Number of directions that have anything dense
+	var/num_dense_found = 0
+	// Number of directions that have 2 squares of open space.
+	var/num_very_open_floors = 0
+	for(var/dir in GLOB.cardinals)
+		var/turf/neighbor_turf = get_step(turf, dir)
+		if(isclosedturf(neighbor_turf))
+			num_walls_found++
+			num_dense_found++
+			continue
+		if(mode == AREA_SPAWN_MODE_HUG_WALL)
+			var/turf/long_test_turf = get_step(neighbor_turf, dir)
+			if(isopenturf(long_test_turf))
+				num_very_open_floors++
+		for(var/atom/movable/found_movable in neighbor_turf)
+			if(istype(found_movable, /obj/effect))
+				continue
+
+			if(found_movable.density || is_type_in_list(found_movable, restricted_objects_list))
+				found_dense_object = TRUE
+				num_dense_found++
+				break
+
+	// Wall hugging also, as a low priority, doesn't even want diagonal things.
+	var/num_diagonal_objects = 0
+	if(mode == AREA_SPAWN_MODE_HUG_WALL)
+		for(var/dir in GLOB.diagonals)
+			var/turf/neighbor_turf = get_step(turf, dir)
+			for(var/atom/movable/found_movable in neighbor_turf)
+				if(istype(found_movable, /obj/effect))
+					continue
+
+				if(
+					!is_type_in_list(found_movable, allowed_diagonal_objects_list) \
+					&& (found_movable.density || is_type_in_list(found_movable, restricted_objects_list))
+				)
+					num_diagonal_objects++
+					break
+
+	switch(mode)
+		if(AREA_SPAWN_MODE_OPEN)
+			// For non-wall hug
+			// #1 priority is totally empty
+			// #2 priority is being in the middle of the room
+			return (totally_empty ? 10 : 0) + (4 - num_dense_found)
+
+		if(AREA_SPAWN_MODE_HUG_WALL)
+			// For wall hugging, must be against wall, and not touching another dense object as it may completely block it.
+			if(num_walls_found == 0 || found_dense_object || num_walls_found == 4)
+				return 0
+
+			// #1 Priority after that: be in a totally empty square
+			// #2 (marginally) have clear diagnals
+			// #3 favor being in a cozy wall nook
+			// #4 be in a big room/hallway so we don't pinch a room down to 1 square of passage.
+			return (totally_empty ? 1000 : 0) + (400 - num_diagonal_objects * 100) + (num_walls_found * 10) + num_very_open_floors
+
+		if(AREA_SPAWN_MODE_MOUNT_WALL)
+			// For mounting to walls. Must be against wall.
+			if(num_walls_found == 0 || num_walls_found == 4)
+				return 0
+
+			// #1 Priority after that: be in a totally empty square
+			// #2, actually don't be in a nook!
+			return (totally_empty ? 10 : 0) + (4 - num_walls_found)
+
+	CRASH("Invalid area spawn mode [mode]!")
```

Comment:
```suggestion
/datum/controller/subsystem/area_spawn/proc/calculate_turf_priority(turf/turf, mode)
	// Only spawn on actual floors
	if(!isfloorturf(turf))
		return 0

	var/totally_empty = TRUE
	for(var/atom/movable/found_movable in turf)
		if(iseffect(found_movable))
			continue

		// Some tile conditions for no-go
		if(mode == AREA_SPAWN_MODE_MOUNT_WALL)
			// Different blacklist logic than normal. See flip_density_wall_mount_objects_list
			var/flip_density = is_type_in_list(found_movable, flip_density_wall_mount_objects_list)
			if(
				found_movable.density != flip_density \
				|| (!flip_density && is_type_in_list(found_movable, restricted_objects_list))
			)
				return 0

			// For wall mounts, we actually don't want to overlap wall items.
			if(found_movable.layer > LOW_OBJ_LAYER)
				totally_empty = FALSE

			continue

		// Every other mode.
		if(
			found_movable.density \
			|| is_type_in_list(found_movable, restricted_objects_list) \
			|| is_type_in_list(found_movable, restricted_overlap_objects_list)
		)
			return 0

		if(found_movable.layer > LOW_OBJ_LAYER && found_movable.layer < ABOVE_MOB_LAYER)
			totally_empty = FALSE

	// Number of directions that have a closed wall
	var/num_walls_found = 0
	// Found a dense object?
	var/found_dense_object = FALSE
	// Number of directions that have anything dense
	var/num_dense_found = 0
	// Number of directions that have 2 squares of open space.
	var/num_very_open_floors = 0
	for(var/dir in GLOB.cardinals)
		var/turf/neighbor_turf = get_step(turf, dir)
		if(isclosedturf(neighbor_turf))
			num_walls_found++
			num_dense_found++
			continue
		if(mode == AREA_SPAWN_MODE_HUG_WALL)
			var/turf/long_test_turf = get_step(neighbor_turf, dir)
			if(isopenturf(long_test_turf))
				num_very_open_floors++
		for(var/atom/movable/found_movable in neighbor_turf)
			if(iseffect(found_movable))
				continue

			if(found_movable.density || is_type_in_list(found_movable, restricted_objects_list))
				found_dense_object = TRUE
				num_dense_found++
				break

	switch(mode)
		if(AREA_SPAWN_MODE_OPEN)
			// For non-wall hug
			// #1 priority is totally empty
			// #2 priority is being in the middle of the room
			return (totally_empty ? 10 : 0) + (4 - num_dense_found)

		if(AREA_SPAWN_MODE_HUG_WALL)
			// Wall hugging also, as a low priority, doesn't even want diagonal things.
			var/num_diagonal_objects = 0
			for(var/dir in GLOB.diagonals)
				var/turf/neighbor_turf = get_step(turf, dir)
				for(var/atom/movable/found_movable in neighbor_turf)
					if(iseffect(found_movable))
						continue

					if(
						!is_type_in_list(found_movable, allowed_diagonal_objects_list) \
						&& (found_movable.density || is_type_in_list(found_movable, restricted_objects_list))
					)
						num_diagonal_objects++
						break

			// For wall hugging, must be against wall, and not touching another dense object as it may completely block it.
			if(found_dense_object || num_walls_found == 0 || num_walls_found == 4)
				return 0

			// #1 Priority after that: be in a totally empty square
			// #2 (marginally) have clear diagnals
			// #3 favor being in a cozy wall nook
			// #4 be in a big room/hallway so we don't pinch a room down to 1 square of passage.
			return (totally_empty ? 1000 : 0) + (400 - num_diagonal_objects * 100) + (num_walls_found * 10) + num_very_open_floors

		if(AREA_SPAWN_MODE_MOUNT_WALL)
			// For mounting to walls. Must be against wall.
			if(num_walls_found == 0 || num_walls_found == 4)
				return 0

			// #1 Priority after that: be in a totally empty square
			// #2, actually don't be in a nook!
			return (totally_empty ? 10 : 0) + (4 - num_walls_found)

	CRASH("Invalid area spawn mode [mode]!")

```